src/main/src/main/java/org/geoserver/security/decorators/DefaultSecureCatalogFactory.java

```diff
@@ -29,32 +37,145 @@ public boolean canSecure(Class clazz) {
+            return new SecuredCoverageStoreInfo(unwrap((CoverageStoreInfo) object), policy);
         else if (DataStoreInfo.class.isAssignableFrom(clazz))
-            return new SecuredDataStoreInfo((DataStoreInfo) object, policy);
+            return new SecuredDataStoreInfo(unwrap((DataStoreInfo) object), policy);
         else if (FeatureTypeInfo.class.isAssignableFrom(clazz))
-            return new SecuredFeatureTypeInfo((FeatureTypeInfo) object, policy);
+            return new SecuredFeatureTypeInfo(unwrap((FeatureTypeInfo) object), policy);
         else if (LayerInfo.class.isAssignableFrom(clazz))
-            return new SecuredLayerInfo((LayerInfo) object, policy);
+            return new SecuredLayerInfo(unwrap((LayerInfo) object), policy);
+        else if (WMTSLayerInfo.class.isAssignableFrom(clazz))
+            return new SecuredWMTSLayerInfo(unwrap((WMTSLayerInfo) object), policy);
         else
-            throw new IllegalArgumentException("Don't know how to wrap");
+            throw new IllegalArgumentException("Don't know how to wrap " + object);
     }
-
     /**
      * Returns {@link ExtensionPriority#LOWEST} since the wrappers generated by
      * this factory
+     * @return {@link ExtensionPriority#LOWEST}
      */
+    @Override
     public int getPriority() {
         return ExtensionPriority.LOWEST;
     }
 
+    private void logDoubleWrap(Object unwrapped, Object orig) {
+        String msg = String.format("Tried to double secure: %s already securing %s", orig, unwrapped);
+        LOGGER.warning(msg);
+    }
+
+    private void logDoubleWrap(final DataStoreInfo store) {
+        int proxies = 0;
+        int secured = 0;
+        DataStoreInfo object = store;
+        while (true) {
+            DataStoreInfo unwrapped;
+            if (object instanceof SecuredDataStoreInfo) {
+                secured++;
+                unwrapped = ((SecuredDataStoreInfo) object).unwrap(DataStoreInfo.class);
+            } else {
+                unwrapped = ModificationProxy.unwrap(object);
+                boolean isProxy = object != unwrapped;
+                if (isProxy) {
+                    proxies++;
+                } else {
+                    break;
+                }
+            }
+            object = unwrapped;
+        }
+
+        String msg = String.format("Double securing %s: proxies: %,d, SecuredDataStoreInfos: %,d",
+                store.getName(), proxies, secured);
+        LOGGER.warning(msg);
+    }
+
+    private WMTSLayerInfo unwrap(WMTSLayerInfo object) {
+        WMTSLayerInfo unwrapped = ModificationProxy.unwrap(object);
+        boolean isProxy = object != unwrapped;
+        if(unwrapped instanceof SecuredWMTSLayerInfo) {
+            logDoubleWrap(unwrapped, object);
+            return unwrap(((SecuredWMTSLayerInfo)unwrapped).unwrap(WMTSLayerInfo.class));
+        }
+        if (isProxy) {
+            object = ModificationProxy.create(unwrapped, WMTSLayerInfo.class);
```

Comment:
Is it possible for the modification proxy to have had modifications made to it? when this happens.  They would be lost in this process.

Reply:
Looking at ModificationProxy more closely, I see your point.

Reply:
@smithkm 
I've retested this without the unwrapping in DefaultSecurityCatalogFactory and my issue seems to still be resolved. I'll leave the logging in, in case a similar issue is hit in the future. It seems the bulk of what I need is in making sure SecuredFeatureStoreInfo *undecorates* DataStoreInfo objects on **setStore()**, to make it the reciprocal of the secure decorating done in **getStore()**. This is consistent with how SecuredLayerInfo handles things here: https://github.com/geoserver/geoserver/blob/master/src/main/src/main/java/org/geoserver/security/decorators/SecuredLayerInfo.java#L30-L58